integration_tests/test_list_01.py

```diff
@@ -42,8 +42,30 @@ def test_list_02():
     for j in range(len(y)):
         assert x[j] == y[j]
 
+# Negative Indexing
+def test_list_03():
+    x: list[f64] = []
+
+    i: i32
+    for i in range(2):
+        x.append(float(i))
+
+    assert x[1] == x[-1]
+    assert x[0] == x[-2]
+    assert x[-1] == 1.0
+    assert x[-2] == 0.0
```

Comment:
So only constant negative indices are allowed? If the same thing is done by a variable then it won't be allowed?

Reply:
Yup, Only Constant negative index works.
If the user pass `variable` which contains the `neg value`, the output value will be `zero`.
But if the user `enables bound check`, then it will be caught as `out of range`.

Reply:
You can do it with a variable, but it has to have a constant value. Using `x[-1]` is very common in Python and I think it's harmless to allow it without losing any performance. Using `x[i]` for negative `i` is  much less common, and we can't easily support it without losing performance. So this is a good compromise I think.

Reply:
It's a good design point. The contract is that if LPython compiles your code *and* there are no runtime errors, then it will behave exactly as in Python. You *have* to use bounds checking in Debug mode.

Whether or not we allow compile time negative values, your example `traverse_list` would still silently fail at runtime if bounds checking is not enabled. This is an unrelated problem. Nothing is stopping a user to call it with negative values.

The same issue is with any runtime indexing, especially arrays. It can be out of bounds and we can't be checking it all the time in Release mode.  The same with integer overflow. We have to be using smaller integers like i32 or i64, but they can overflow. There is nothing we can do about it, and we should not be checking it all the time, it will be slow. But in Debug mode I think we can check it reasonably fast.

We can have a ReleaseSafe mode which keeps all these runtime checks in place, but otherwise optimizes everything out as much as it can. In this mode, it would indeed not fail silently. But the ReleaseSafe mode will almost for sure be slower than just Release mode.

Reply:
> We can have a ReleaseSafe mode which keeps all these runtime checks in place, but otherwise optimizes everything out as much as it can. In this mode, it would indeed not fail silently. But the ReleaseSafe mode will almost for sure be slower than just Release mode.

This is also okay for me. At least we should provide a release which is as safe as possible and gives runtime errors/warnings for issues we think are possible. We can specify at the top of our docs that we provide following releases (ReleaseSafe, Production, Debug) and then users will be responsible for their choice.

Reply:
We can call it Debug, ReleaseSafe, ReleaseFast. 

Both of the releases will be used in production. Webbrowser will use ReleaseSafe, an HPC computational code would use ReleaseFast.

Also, if webassembly is safe (?), then we could use ReleaseFast with a WASM output as well. It just depends on what you want to do.

Reply:
https://github.com/lcompilers/lpython/issues/1139